lib/r10k/hg/stateful_repository.rb

```diff
@@ -0,0 +1,86 @@
+require 'r10k/hg/cache'
+require 'r10k/hg/errors'
+require 'r10k/hg/working_repository'
+require 'forwardable'
+require 'r10k/logging'
+
+# Manage how Mecurial repositories are created and set to specific revisions
+class R10K::Hg::StatefulRepository
+
+  include R10K::Logging
+
+  # @!attribute [r] repo
+  #   @api private
+  attr_reader :repo
+
+  extend Forwardable
+  def_delegators :@repo, :head
+
+  # Create a new Mecurial working directory
+  #
+  # @param rev     [String] The Mecurial revision to check out
+  # @param remote  [String] The Mecurial remote to use for the repo
+  # @param basedir [String] The path containing the Mecurial repo
+  # @param dirname [String] The directory name of the Mecurial repo
+  def initialize(rev, remote, basedir, dirname)
+    @rev = rev
+    @remote = remote
+
+    @cache = R10K::Hg::Cache.generate(remote)
+    @repo = R10K::Hg::WorkingRepository.new(basedir, dirname)
+  end
+
+  def sync
+    @cache.sync if sync_cache?
+
+    sha = @cache.resolve(@rev)
+
+    if sha.nil?
+      raise R10K::Hg::UnknownRevisionError.new("Unable to sync repo to unresolvable revision '#{@rev}'", :hg_dir => @repo.hg_dir)
+    end
+
+    case status
+      when :absent
+        logger.debug { "Cloning #{@repo.path} and checking out #{@rev}" }
+        @repo.clone(@cache.hg_dir.to_s, {:rev => sha})
```

Comment:
I was digging through the Mercurial man pages just now looking for a way to improve this. One easy improvement is passing the `--branch` parameter when calling `clone` and `pull` for the environment clones. That way we only clone/pull the data that is relevant for each environment rather than cloning everything for each environment.